dali/kernels/normalize/normalize_gpu_impl.cuh

```diff
@@ -0,0 +1,575 @@
+           float base, const InListGPU<Scale> &scale,
+           float global_scale = 1, float shift = 0, float epsilon = 0) {
+    CheckDataShape(out.shape, in.shape);
+
+    if (!scalar_base_ || scalar_scale_)
+      throw std::logic_error("Normalize was not set up for use with scalar base.");
+
+    CheckParamShape(in.shape, scale.shape);
+
+    if (scale_is_stddev_)
+      RunInvStdDev<Op_InvStdDevScalarBase>(ctx, out, in, base, scale, epsilon, global_scale, shift);
+    else
+      RunScale<Op_ScalarBase>(ctx, out, in, base, scale, global_scale, shift);
+  }
+
+  void Run(KernelContext &ctx,
+           const OutListGPU<Out> &out, const InListGPU<In> &in,
+           float base, float scale,
+           float global_scale = 1, float shift = 0, float epsilon = 0) {
+    CheckDataShape(out.shape, in.shape);
+
+    if (!scalar_base_ || !scalar_scale_)
+      throw std::logic_error("Normalize was not set up for use with scalar arguments.");
+
+    if (scale_is_stddev_)
+      scale = epsilon ? rsqrt(scale * scale + epsilon) : 1 / scale;
+    RunScale<Op_Scalar>(ctx, out, in, base, scale, global_scale, shift);
+  }
+
+ private:
+  void CheckParamShape(const TensorListShape<> &data_shape, const TensorListShape<> &param_shape) {
+    bool broadcast_param = param_shape.num_samples() == 1;
+    int D = data_shape.sample_dim();
+    int N = data_shape.num_samples();
+    for (int i = 0, p = 0; i < N; i++, p += !broadcast_param) {
+      auto dshape = data_shape[i];
+      auto pshape = param_shape[p];
+      for (int d = 0; d < D; d++) {
+        if (axis_mask_ & (1 << d)) {
+          if (pshape[d] != 1) {
+            throw std::invalid_argument(make_string("Parameter tensor must have extent 1 "
+              "in reduced axes. Got:"
+              "\nparameter shape: ", pshape,
+              "\nreduced axes:    ", axes_str()));
+          }
+        } else {
+          if (pshape[d] != dshape[d]) {
+            throw std::invalid_argument(make_string("Parameter tensor must have extent "
+              "equal to input shape in non reduced axes. Got:"
+              "\nparameter shape: ", pshape,
+              "\ndata shape: ", dshape,
+              "\nreduced axes:    ", axes_str()));
+          }
+        }
+      }
+    }
+  }
+
+  void CheckDataShape(const TensorListShape<> &out_shape, const TensorListShape<> &in_shape) {
+    if (out_shape != in_shape)
+      throw std::invalid_argument("Output and input must have the same shape");
+    if (in_shape.sample_dim() != ndim_)
+      throw std::invalid_argument("The input tensor list has different dimensionality than the "
+        "shape passed to Setup");
+    if (in_shape.num_samples() != num_samples_)
+      throw std::invalid_argument("The input tensor list has different number of samples than the "
+        "shape passed to Setup");
+  }
+
+  std::pair<dim3, dim3> GetLaunchParams(const TensorListShape<> &data_shape) const {
+    int64_t block = 1024;
```

Comment:
Maybe we should query the GPU for capabilities - I mean user should and provide the necessary values to the kernel?